### DISCUSSION TITLE
Async code/request scope?

### DISCUSSION THREAD

**Discussion by chbndrhnns** [illuin-tech/opyoid#26] [Q&A]
Is there a way to make opyoid work with web frameworks like [FastAPI](https://fastapi.tiangolo.com/) which work asynchronously?

I image a scope which is valid for one request would be really helpful here, probably using `contextvars`.
For example, the `dependency_injector` library uses a `ContextLocalSingleton` scope.

**Comment by Aegdesil** [comment]
FastAPI uses [its own dependency injection mecanism](https://fastapi.tiangolo.com/tutorial/dependencies/#what-is-dependency-injection), why not use it ?
If your question was more about the possibility to make custom scopes, you can check [this documentation](https://github.com/illuin-tech/opyoid/blob/master/docs/custom_scopes.md) to learn more about custom scopes, you could use this to create a request-scoped context.

**Reply by chbndrhnns** [reply]
Thanks for the reply! I find the FastAPI DI system really convenient for simple applications but it does not help me when structuring my app like in the hexagonal architecture where the web framework is not the center of everyting.
Besides that, the DI system of FastAPI cannot be used outside of a request context, e.g. during startup or shutdown.
